File: apps/helper-go/internal/mikan/rss.go
package mikan

import (
	"regexp"
	"strconv"
	"strings"
)

type RssEpisode struct {
	EpisodeID   string `json:"episodeId"`
	Title       string `json:"title"`
	TorrentURL  string `json:"torrentUrl"`
	PublishedAt string `json:"publishedAt,omitempty"`
	SizeBytes   int64  `json:"sizeBytes,omitempty"`
}

var (
	reItem        = regexp.MustCompile(`(?is)<item\b[^>]*>.*?</item>`)
	reEnclosure   = regexp.MustCompile(`(?i)<enclosure\b[^>]*>`)
	reEpisodePage = regexp.MustCompile(`(?i)/Home/Episode/([a-f0-9]+)`)
	reEpisodeDL   = regexp.MustCompile(`(?i)/Download/\d+/([a-f0-9]+)\.torrent`)
	reHexEntity   = regexp.MustCompile(`&#x([0-9a-fA-F]+);`)
	reDecEntity   = regexp.MustCompile(`&#(\d+);`)
)

func ParseBangumiRSS(xml, baseURL string) []RssEpisode {
	items := reItem.FindAllString(xml, -1)
	episodes := make([]RssEpisode, 0, len(items))
	for _, item := range items {
		title := tagText(item, "title")
		pageLink := tagText(item, "link")
		enclosureTag := ""
		if match := reEnclosure.FindString(item); match != "" {
			enclosureTag = match
		}
		enclosureURL := attr(enclosureTag, "url")
		episodeID := ""
		if match := reEpisodePage.FindStringSubmatch(pageLink); match != nil {
			episodeID = match[1]
		} else if match := reEpisodeDL.FindStringSubmatch(enclosureURL); match != nil {
			episodeID = match[1]
		}
		torrentHref := enclosureURL
		if torrentHref == "" {
			torrentHref = pageLink
		}
		if title == "" || episodeID == "" || torrentHref == "" {
			continue
		}
		episode := RssEpisode{
			EpisodeID:  episodeID,
			Title:      title,
			TorrentURL: TorrentDownloadURL(baseURL, torrentHref),
		}
		if published := tagText(item, "pubDate"); published != "" {
			episode.PublishedAt = published
		}
		sizeRaw := tagText(item, "contentLength")
		if sizeRaw == "" {
			sizeRaw = attr(enclosureTag, "length")
		}
		if size, err := strconv.ParseInt(sizeRaw, 10, 64); err == nil && size > 0 {
			episode.SizeBytes = size
		}
		episodes = append(episodes, episode)
	}
	return episodes
}

func tagText(xml, name string) string {
	re := regexp.MustCompile(`(?is)<` + regexp.QuoteMeta(name) + `(?:\s[^>]*)?>(.*?)</` + regexp.QuoteMeta(name) + `>`)
	match := re.FindStringSubmatch(xml)
	if match == nil {
		return ""
	}
	return decodeXML(match[1])
}

func attr(tag, name string) string {
	if tag == "" {
		return ""
	}
	re := regexp.MustCompile(`(?i)` + regexp.QuoteMeta(name) + `\s*=\s*"([^"]*)"`)
	if match := re.FindStringSubmatch(tag); match != nil {
		return decodeXML(match[1])
	}
	re = regexp.MustCompile(`(?i)` + regexp.QuoteMeta(name) + `\s*=\s*'([^']*)'`)
	if match := re.FindStringSubmatch(tag); match != nil {
		return decodeXML(match[1])
	}
	return ""
}

func decodeXML(value string) string {
	out := reHexEntity.ReplaceAllStringFunc(value, func(match string) string {
		sub := reHexEntity.FindStringSubmatch(match)
		n, err := strconv.ParseInt(sub[1], 16, 32)
		if err != nil {
			return match
		}
		return string(rune(n))
	})
	out = reDecEntity.ReplaceAllStringFunc(out, func(match string) string {
		sub := reDecEntity.FindStringSubmatch(match)
		n, err := strconv.ParseInt(sub[1], 10, 32)
		if err != nil {
			return match
		}
		return string(rune(n))
	})
	out = strings.ReplaceAll(out, "&nbsp;", " ")
	out = strings.ReplaceAll(out, "&quot;", `"`)
	out = strings.ReplaceAll(out, "&apos;", "'")
	out = strings.ReplaceAll(out, "&lt;", "<")
	out = strings.ReplaceAll(out, "&gt;", ">")
	out = strings.ReplaceAll(out, "&amp;", "&")
	return out
}

File: packages/mikan/src/rss.ts
import { torrentDownloadUrl } from './urls'

export interface RssEpisode {
  episodeId: string
  publishedAt?: string
  sizeBytes?: number
  title: string
  torrentUrl: string
}

function decodeXml(value: string): string {
  return value
    .replaceAll(/&#x([\dA-Fa-f]+);/g, (_, hex: string) =>
      String.fromCodePoint(Number.parseInt(hex, 16)),
    )
    .replaceAll(/&#(\d+);/g, (_, dec: string) =>
      String.fromCodePoint(Number.parseInt(dec, 10)),
    )
    .replaceAll('&nbsp;', ' ')
    .replaceAll('&quot;', '"')
    .replaceAll('&apos;', "'")
    .replaceAll('&lt;', '<')
    .replaceAll('&gt;', '>')
    .replaceAll('&amp;', '&')
}

function tagText(xml: string, name: string): string | undefined {
  const match = xml.match(
    new RegExp(`<${name}(?:\\s[^>]*)?>([\\s\\S]*?)</${name}>`, 'i'),
  )
  return match?.[1] === undefined ? undefined : decodeXml(match[1])
}

function attr(tag: string, name: string): string | undefined {
  const match =
    tag.match(new RegExp(`${name}\\s*=\\s*"([^"]*)"`, 'i')) ??
    tag.match(new RegExp(`${name}\\s*=\\s*'([^']*)'`, 'i'))
  return match?.[1] === undefined ? undefined : decodeXml(match[1])
}

export function parseBangumiRss(xml: string, baseUrl: string): RssEpisode[] {
  const items = xml.match(/<item\b[^>]*>[\S\s]*?<\/item>/gi) ?? []
  const episodes: RssEpisode[] = []

  for (const item of items) {
    const title = tagText(item, 'title')
    const pageLink = tagText(item, 'link') ?? ''
    const enclosureTag = item.match(/<enclosure\b[^>]*>/i)?.[0] ?? ''
    const enclosureUrl = attr(enclosureTag, 'url')
    const episodeId =
      pageLink.match(/\/home\/episode\/([\da-f]+)/i)?.[1] ??
      enclosureUrl?.match(/\/download\/\d+\/([\da-f]+)\.torrent/i)?.[1]
    const torrentHref = enclosureUrl ?? pageLink
    if (!title || !episodeId || !torrentHref) {
      continue
    }

    const episode: RssEpisode = {
      episodeId,
      title,
      torrentUrl: torrentDownloadUrl(baseUrl, torrentHref),
    }

    const publishedAt = tagText(item, 'pubDate')
    if (publishedAt) {
      episode.publishedAt = publishedAt
    }

    const sizeRaw =
      tagText(item, 'contentLength') ?? attr(enclosureTag, 'length')
    const sizeBytes = Number(sizeRaw)
    if (Number.isFinite(sizeBytes) && sizeBytes > 0) {
      episode.sizeBytes = sizeBytes
    }

    episodes.push(episode)
  }

  return episodes
}
